### PR TITLE
Snapdragon Spektrum input

### PR DESCRIPTION
**Pull request by LorenzMeier**
@Stifael @julianoes @bkueng This is the rough structure (it is *not* compiling) which enables Spektrum RC input on Snapdragon. The code is taken from here:
https://github.com/PX4/Firmware/blob/master/src/drivers/px4fmu/fmu.cpp#L1342-L1354

And you will find anything missing there.

@bharath374 @jywilson @mcharleb I would appreciate if you could help pushing this to completion. It hurts us quite a bit that this is not an open source driver and we really should get onto one consistently maintained platform as the user experience will continue to be horrible until we are there.

### CODE REVIEW

**Comment by julianoes** [comment]
@Stifael please review and test :)

**Comment by LorenzMeier** [comment]
Jenkins test this please.

File: src/lib/rc/dsm.c [hunk 1]
```diff
@@ -305,6 +305,13 @@ dsm_init(const char *device)
 	}
 }
 
+void
+dsm_deinit()
+{
+	close(dsm_fd);
```
Comment (LorenzMeier) [inline src/lib/rc/dsm.c]: Should check if FD is >= 0.

Comment (julianoes) [inline src/lib/rc/dsm.c]: Done in 871d821

**Comment by LorenzMeier** [comment]
@Stifael Please review & test.

File: src/drivers/spektrum_rc/spektrum_rc.cpp [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+		}
+
+		if (!(fds[0].revents & POLLIN)) {
+			// There is only one fd in poll, so this should be impossible.
+			PX4_WARN("poll bit not set");
+			continue;
+		}
+
+		int newbytes = ::read(uart_fd, &rx_buf[0], sizeof(rx_buf));
+
```
Comment (Stifael) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: being pick:
can write: 
::read(uart_fd, rx_buf, sizeof(rx_buf))

File: src/drivers/spektrum_rc/spektrum_rc.cpp [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+		const hrt_abstime now = hrt_absolute_time();
+
+		bool dsm_11_bit;
+		unsigned frame_drops;
+		uint16_t raw_rc_values[input_rc_s::RC_INPUT_MAX_CHANNELS];
+		uint16_t raw_rc_count;
+
+		// parse new data
+		bool rc_updated = dsm_parse(now, rx_buf, newbytes, &raw_rc_values[0], &raw_rc_count,
+					    &dsm_11_bit, &frame_drops, input_rc_s::RC_INPUT_MAX_CHANNELS);
```
Comment (Stifael) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: being picky:
can write "raw_rc_values"

File: src/drivers/spektrum_rc/spektrum_rc.cpp [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+	if (valid_chans == 0) {
+		input_rc.rssi = 0;
+	}
+
+	input_rc.rc_failsafe = failsafe;
+	input_rc.rc_lost = (valid_chans == 0);
+	input_rc.rc_lost_frame_count = frame_drops;
+	input_rc.rc_total_frame_count = 0;
+}
```
Comment (Stifael) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: why exactly are rc_ppm_frame_length and rc_total_frame_count set to 0?

File: src/drivers/spektrum_rc/spektrum_rc.cpp (hunk 2 above)
Comment (LorenzMeier) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: The current notion is actually better. Using the more verbose notation makes obvious that this is accessing the first position of the array.

File: src/drivers/spektrum_rc/spektrum_rc.cpp (hunk 4 above)
Comment (julianoes) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: `rc_ppm_frame_lengh`:
https://github.com/PX4/Firmware/blob/master/msg/input_rc.msg#L25
0 for non-PPM systems

`rc_total_frame_count`:
https://github.com/PX4/Firmware/blob/master/msg/input_rc.msg#L24
We should just increment it but I don't really see the value of this field.

File: src/drivers/spektrum_rc/spektrum_rc.cpp (hunk 3 above)
Comment (LorenzMeier) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: Same here. This notion is more explicit and more prone to result in (correct) compile errors if the variable would not hold an array but something else.

File: src/drivers/spektrum_rc/spektrum_rc.cpp (hunk 4 above)
Comment (LorenzMeier) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: Because we don't receive these two.

File: src/drivers/spektrum_rc/spektrum_rc.cpp (hunk 2 above)
Comment (julianoes) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: fixed in ef3bc14

File: src/drivers/spektrum_rc/spektrum_rc.cpp (hunk 3 above)
Comment (julianoes) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: fixed in ef3bc14

**Comment by LorenzMeier** [comment]
@julianoes Please remove ef3bc14. The rationale is given here:
http://www.drdobbs.com/cpp/still-be-explicit-to-avoid-those-arraypo/228701363

  * You can't replace the array with a std::vector
  * Some compilers do not support decaying the array to a pointer

**Comment by julianoes** [comment]
Reverted ef3bc14.

**Comment by Stifael** [comment]
ok, good to know and will from now on use the explicit version. 
I just tried to compile it on snapdragon, and there is an issue with the new function (dsm_init, dsm_deinit, dsm_parse from dsm.c)  which are not found by libpx4.so.

**Comment by LorenzMeier** [comment]
We need to add src/lib/rc/dsm to the compilation @julianoes.

**Comment by jywilson** [comment]
Please use the aDSP static image available here:
https://www.intrinsyc.com/fchsdkaddon/

Intrinsyc is supposed to release this aDSP static image in their next production release, which may be newer than the download this link.  If their production release is older, then use the aDSP image at the link above in the interim.

Jim W.

**Comment by julianoes** [comment]
@jywilson thanks for the image, we'll try it out.

**Comment by julianoes** [comment]
Yep, it runs for me now. @Stifael you can use the image at https://www.intrinsyc.com/fchsdkaddon/ and try connecting a receiver.

**Comment by Stifael** [comment]
@jywilson: I am still trying to receive data, but it does not read anything. I used port tty-1 (that should map to J15 next to usb), and use pin 3 as input, but don't receive any data. I have tried other ports as well, without success. I attached a logic analyzer to see if data is even sent, and that seems correct. Therefore, I am wondering if I use the wrong pin?

File: src/drivers/spektrum_rc/spektrum_rc.cpp [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+	orb_advert_t rc_pub = nullptr;
+
+	// Use a buffer size of the double of the minimum, just to be safe.
+	uint8_t rx_buf[2 * DSM_BUFFER_SIZE];
+
+	// Set up poll topic
+	px4_pollfd_struct_t fds[1];
+	fds[0].fd     = uart_fd;
+	fds[0].events = POLLIN;
+
```
Comment (Stifael) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: Dont use poll (px4_poll) in qurt

File: src/drivers/spektrum_rc/spektrum_rc.cpp [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+			usleep(100000);
+			continue;
+		}
+
+		if (!(fds[0].revents & POLLIN)) {
+			// There is only one fd in poll, so this should be impossible.
+			PX4_WARN("poll bit not set");
+			continue;
+		}
+
```
Comment (Stifael) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: dont use poll (like in the gps driver)

**Comment by Stifael** [comment]
ok, the driver should work. it just requires the deletion of the poll

**Comment by LorenzMeier** [comment]
@Stifael Please push the fix directly to this branch instead of commenting. That's way easier. You also can't just delete poll because that makes the driver run in a tight loop (loading a CPU 100%). You need to find a way to let it either wait for new data (using poll()) or delay it with usleep() to run only at the RC input rate (50 Hz?).

**Comment by jywilson** [comment]
Just using @jywilson should be enough.  No need to use @jywilson2.

Regarding the status of poll in QuRT...

There is no native implementation of poll available in the aDSP.  I have created a ticket to work on this in the future, though I can't tell you the exact date.

The implementation of px4_poll() currently supports QuRT (aDSP).

**Comment by julianoes** [comment]
@Stifael it says you "requested changes" but I think that's done.

File: src/drivers/spektrum_rc/spektrum_rc.cpp [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+		usleep(10000);
+		int newbytes = ::read(uart_fd, &rx_buf[0], sizeof(rx_buf));
+
+		if (newbytes < 0) {
+			PX4_WARN("read failed");
+			continue;
+		}
+
+		if (newbytes == 0) {
+			PX4_WARN("read nothing");
```
Comment (julianoes) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: @Stifael You said this happens fairly often, so I'd remove the `PX4_WARN`. We don't want to spam the console and logs (`PX4_WARN`) is also logged in logger.

File: src/drivers/spektrum_rc/spektrum_rc.cpp [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+	// Use a buffer size of the double of the minimum, just to be safe.
+	uint8_t rx_buf[2 * DSM_BUFFER_SIZE];
+
+	_is_running = true;
+
+	// Main loop
+	while (!_task_should_exit) {
+
+		// sleep since no poll
+		usleep(10000);
```
Comment (julianoes) [inline src/drivers/spektrum_rc/spektrum_rc.cpp]: A bit more intuitive would be to read first and sleep at the end of a cycle.

**Comment by julianoes** [comment]
I should have remembered that poll is not (yet) available, see https://github.com/PX4/DriverFramework/issues/26.

**Comment by Stifael** [comment]
why does it say the build failded on semaphore, but if pressed on details nothing shows up
?

**Comment by LorenzMeier** [comment]
The PR is still incomplete as the driver is not auto-started. But I guess we will do a follow-up when flight tests have been done.

**Comment by LorenzMeier** [comment]
@jywilson Would it be feasible to integrate with you together? We have now RC input and PWM output implemented and it would be very helpful to have someone on your side co-drive it. @Stifael @ChristophTobler please keep us updated on your progress.

**Comment by Stifael** [comment]
Christoph finished yesterday a new snapdragon support and I tested today successful the rc interface. I will now  need to test gps and pwm output, after that we should be ready to fly.

**Comment by LorenzMeier** [comment]
Note that I've rebased the branch, so you need to reset it locally. Talk to Julian or Beat how to do that.

**Comment by Stifael** [comment]
I already did that

**Comment by LorenzMeier** [comment]
This is on master now as well.

**Comment by jywilson** [comment]
Wow this is great news Stifael!  I will also try to repro your results by testing on my own vehicle.  I may not be able to get to it until I get back from vacation in January.  I will keep you posted.

**Pull request merged**